Dev

ligianicolovici opened the pull request:
Please review my chapter 3 implementation in online shop.
Tasks done:
- display one product
- display a list of products
- refactoring and restructuring code

Commit:
create project online shop

Commit:
display products

Commit:
product details and product list dispay

Commit:
refactoring

Commit:
chapter 4 done

Commit:
chapter 5 done

serban-petrescu commented on online-shop/src/App.tsx:
```diff
@@ -0,0 +1,67 @@
+import { faStroopwafel } from '@fortawesome/free-solid-svg-icons';
+library.add(faStroopwafel)
+
+export interface IProduct{
```
I generally recommend moving interfaces / classes which describe business data in a separate, `model` folder. In the `tsx` files, you should just have the component itself + props / state definition + any local helper functions.

serban-petrescu commented on online-shop/src/App.tsx:
```diff
@@ -0,0 +1,67 @@
+  id:number;
+  name:string;
+  category:string;
+  image:string;
```
Also, please use the code formatter built-in your IDE.

serban-petrescu commented on online-shop/src/App.tsx:
```diff
@@ -0,0 +1,67 @@
+  itemsCart:IProduct[];
+}
+export default class App extends React.Component<IProps,IState>{
+   products: IProduct[] = Object.values(PostData);
```
You should avoid using properties in React components. Instead, use the state to store the loaded products.

serban-petrescu commented on online-shop/src/App.tsx:
```diff
@@ -0,0 +1,67 @@
+      itemsCart:[]
+    }
+  }
+  changeStateArray(cartProducts:IProduct[]){
```
The naming of this method could be better, for example: `addProductsToCart`. You should always strive to give meaningful names. https://medium.com/coding-skills/clean-code-101-meaningful-names-and-functions-bf450456d90c

serban-petrescu commented on online-shop/src/components/NavBar.tsx:
```diff
@@ -0,0 +1,48 @@
+import { library, icon } from '@fortawesome/fontawesome-svg-core'
+import { faCamera,faShoppingCart } from '@fortawesome/free-solid-svg-icons'
+
+library.add(faCamera,faShoppingCart)
```
To not lose track of what you added and where, I would suggest having a `icons.ts` file next to your `index.ts`, do all the "add" calls there and then import that file in the `index.ts`.

serban-petrescu commented on online-shop/src/components/ProductDetails.tsx:
```diff
@@ -0,0 +1,82 @@
+export interface IProps {
+    data?:IProduct;
+    match?:any;
+    appSetArray?:any;
```
This should be typed as function: `appSetArray`: (products: IProduct[]) => void.

Also, you can find a better name for it. E.g. onAddProduct or onAddButtonClick. Usually callbacks which are linked to events start with `on`.

Also, you don't need to pass an array. This component should just be able to add one more product to the shoppin cart - it should be enough to pass just the product itself.

serban-petrescu commented on online-shop/src/components/ProductDetails.tsx:
```diff
@@ -0,0 +1,82 @@
+}
+
+
+const products: IProduct[] = Object.values(PostData);
```
Don't use static variables! Use state or props instead.

serban-petrescu commented on online-shop/src/components/ProductList.tsx:
```diff
@@ -0,0 +1,36 @@
+    render() {
+        let products= this.props.data.map(
+            (productt) => 
+                <Link to= {`/products/${productt.id}`}>     
```
Always assign keys to objects created in loops (like `.map` calls). If you check your console, you get for sure an error / warning because you omitted the keys.

serban-petrescu commented on online-shop/src/components/ShoppingCart.tsx:
```diff
@@ -0,0 +1,68 @@
+    }
+    render() {
+        if(this.state.cartProducts!=null){
+              this.productsFromCart= this.state.cartProducts.map(
```
Simply use a local variable instead of a class attribute for the `productsFromCart`.

Commit:
refactor in code after the code review + chapter 6 done except for the checkOut task

Commit:
chapter 6 done

Commit:
chapter 7 done

Commit:
chapter 8 done

serban-petrescu commented on online-shop/src/App.tsx:
```diff
@@ -1,67 +1,62 @@
+            <Route
+              path="/"
+              exact
+              render={props => <ProductList {...props} />}
```
Do you really need to pass these props to the child component?

serban-petrescu commented on online-shop/src/actions/editActions.tsx:
```diff
@@ -0,0 +1,97 @@
+    productList: listOfProducts
+  };
+}
+export function addProduct(
```
Talk to Alex about how to improve your action / reducer structure.

serban-petrescu commented on online-shop/src/components/DetailsManagement.tsx:
```diff
@@ -0,0 +1,297 @@
+              <div className="field">
+                <p className="control is-expanded has-icons-left">
+                  <input
+                    className="input is-primary"
```
This component is huge - you should split it up in smaller components.

serban-petrescu replied:
Also, this component is neither presentational (dumb) nor container (smart). Split it up into smart / dumb components.

serban-petrescu commented on online-shop/src/components/DetailsManagement.tsx:
```diff
@@ -0,0 +1,297 @@
+  fetchUpdateProduct: (id: number, productUpdated: IProduct) => void;
+}
+class DetailsManagement extends React.Component<EditDetailsProps> {
+  categories: string[] = this.getCategories(this.props.productList);
```
Never use class attributes. Instead use state / store / props.

serban-petrescu commented on online-shop/src/components/DetailsManagement.tsx:
```diff
@@ -0,0 +1,297 @@
+    );
+  }
+  getCategories(this: any, data: IProduct[]) {
+    let i: number;
```
I don't get what this does :) Contact me.

serban-petrescu commented on online-shop/src/components/DetailsManagement.tsx:
```diff
@@ -0,0 +1,297 @@
+  mapStateToProps,
+  mapDispatchToProps
+)(DetailsManagement);
+export default EditDetailsInitializer;
```
This should be in its own file.

serban-petrescu commented on online-shop/src/components/ProductDetails.tsx:
```diff
@@ -1,82 +1,213 @@
+  crtCart: state.shoppingReducer.cartProducts
+});
 
+const mapDispatchToProps = (dispatch: Dispatch) => ({
```
Please format your code.

serban-petrescu commented on online-shop/src/reducers/combine.tsx:
```diff
@@ -0,0 +1,21 @@
+import { productEditManipulation } from "./edit";
+
+export interface AppState {
+  productListReducer: ProductListState;
```
It is not the best that your state has properties with the name `blaReducer`. It should be like:

```
{
 products: ...
 productDetails: ...
 shoppingCart: ...
 editProduct: ...
}

```

serban-petrescu commented on online-shop/src/sagas/index.tsx:
```diff
@@ -0,0 +1,119 @@
+}
+
+function* actionWatcher() {
+  yield takeLatest("FETCH-PRODUCTS", fetchProductList);
```
In general, your action names are slightly confusing. Also, you should use constants / enums instead of simple strings.

serban-petrescu commented on online-shop/src/reducers/shopping.tsx:
```diff
@@ -0,0 +1,204 @@
+    if (crtCartItems[i].product.id === newCartProduct.id) {
+      contor = crtCartItems[i].quantity;
+      contor++;
+      crtCartItems[i].quantity = contor;
```
You should avoid modifying the old state, instead you should create a new state object. It would be a lot easier if you stored the products in a map / object (and not an array).

serban-petrescu reviewed (commented):
Overall, your components are too big, also the split between smart / dumb components is not done. You should have tens of dumb (functional) components and just a few or so smart components.

Commit:
components splited in smart and dumb;
elimated class attibutes;
highcharts done;
actions  names changed;
and a few hocs added

Commit:
-navBar component splited in smart and dumb ;
-loading indicator displayed using HOOKS,
-renamed state properties from combine (as recommended)

Commit:
HOCs chapter done

serban-petrescu commented on online-shop/src/components/Details/smartDetails.tsx:
```diff
@@ -0,0 +1,163 @@
+  isLoading:boolean;
+  switchLoading:()=>void;
+}
+const onComponentDidMount = lifecycle<ProductDetailsProps, {}, {}>({
```
How come this is outside of the class defined below? Using `lifecycle` makes sense just when you don't have an underlying class (but a functional component - a dumb component).

serban-petrescu commented on online-shop/src/components/HOCS/LoaderHOC.tsx:
```diff
@@ -0,0 +1,12 @@
+  class WithLoading extends React.Component<P & LoadingProps> {
+    render() {
+        console.log(this.props.isLoading);
+      return  this.props.isLoading ? <div className="pageloader is-active"><span className="title">Loading</span></div>: <Component {...this.props as P} />;
```
Typically you would not pass the loading indicator anymore to the underlying component - because it will anyway be always false when the component renders (due to the ternary condition).

serban-petrescu commented on online-shop/src/sagas/index.tsx:
```diff
@@ -1,29 +1,56 @@
 function* fetchProductList() {
   const data = yield fetch("http://localhost:4000/products", { method: "GET" })
     .then(response => response.json())
     .then(result => {
```
This does nothing :) did you use it to debug?

serban-petrescu commented on online-shop/src/sagas/index.tsx:
```diff
@@ -1,29 +1,56 @@
 }
 
 function* fetchDeleteDetailsProduct(action: FetchSelectedProduct) {
   yield fetch("http://localhost:4000/products/" + action.productID, {
```
I would recommend extracting the fetch calls (+ corresponding then / catch) into service / API objects in a separate folder and just call these service methods here in the sagas. That way the saga looks cleaner.

serban-petrescu reviewed (commented):
Good job! Some components are still a little too big + we could improve the naming of some things, but overall your code quality improved.
